iceberg/rest_client: use /v1/config endpoint

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
Changes the iceberg_rest_catalog_prefix to
iceberg_rest_catalog_warehouse, with the expectation that this more
closely aligns with pyiceberg and other tools that allow configuring the
"warehouse".

This works by sending a request to the /v1/config endpoint with the
warehouse as a query parameter, and using the 'prefix' field returned by
the catalog, if any.

If no prefix is returned and a warehouse was specified, the warehouse
will used as a prefix instead as a fallback.

I've updated existing tests to use the new iceberg_rest_catalog_warehouse
instead of prefix. I manually verified this against a Unity Catalog, whose
prefix was returned as "catalogs/warehouse_name".

More info about the config endpoint:

- REST OpenAPI spec https://github.com/apache/iceberg/blob/9199ab520b92084e058d7cf7101b15802eeefc7c/open-api/rest-catalog-open-api.yaml#L81-L102
- PyIceberg fetch config implementation https://github.com/apache/iceberg-python/blob/96e6d546f1ac017781bb94c5370f869cae31bd4c/pyiceberg/catalog/rest.py#L376-L394

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* The cluster property `iceberg_rest_catalog_prefix` is renamed to `iceberg_rest_catalog_warehouse` and will now use the given value to query the `/v1/config` endpoint to discover the appropriate REST prefix.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
iceberg/rest_client: use /v1/config endpoint

Changes the iceberg_rest_catalog_prefix to
iceberg_rest_catalog_warehouse, with the expectation that this more
closely aligns with pyiceberg and other tools that allow configuring the
"warehouse".

This works by sending a request to the /v1/config endpoint with the
warehouse as a query parameter, and using the 'prefix' field returned by
the catalog, if any.

If no prefix is returned and a warehouse was specified, the warehouse
will used as a prefix instead as a fallback.

I've updated existing tests to use the new
iceberg_rest_catalog_warehouse instead of prefix. I also manually
verified this against a Unity Catalog, whose prefix was returned as
"catalogs/<warehouse>".

WillemKauf commented on src/v/datalake/coordinator/catalog_factory.cc:
```diff
@@ -205,28 +205,27 @@ rest_catalog_factory::create_catalog() {
       std::move(creds_and_token.credentials),
       std::move(endpoint_information.base_path),           // base_path
-      std::move(prefix_path),                              // prefix
+      warehouse,                                           // warehouse
```
nit: `std::move()`?

andrwng replied:
Done

WillemKauf commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -104,12 +105,74 @@ catalog_client::catalog_client(
+                    .and_then(parse_json)
+                    .and_then(
+                      parse_as_expected("get_config", parse_catalog_config));
+    auto prefix_prop_name = "prefix";
```
nit: `static`?

andrwng replied:
Done

andrwng commented:
More info about the config endpoint:
- REST OpenAPI spec https://github.com/apache/iceberg/blob/9199ab520b92084e058d7cf7101b15802eeefc7c/open-api/rest-catalog-open-api.yaml#L81-L102
- PyIceberg fetch config implementation https://github.com/apache/iceberg-python/blob/96e6d546f1ac017781bb94c5370f869cae31bd4c/pyiceberg/catalog/rest.py#L376-L394

WillemKauf commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -104,12 +105,74 @@ catalog_client::catalog_client(
+        prefix = prefix_path{*_warehouse};
+    }
+    // TODO: use the config for more than just prefix setting.
+    _path_components.reset_prefix(std::move(prefix));
```
can `prefix` still be `std::nullopt` here for any reason? I guess `trim_slashes()` saves us from `_prefix` not being value-less by the time we hit the `vassert()` check in `root_path()`?

andrwng replied:
We should be guaranteed to have it set, even if it's an empty string. Left a comment around reset_prefix()'s definition

WillemKauf commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -104,12 +105,74 @@ catalog_client::catalog_client(
+    if (config) {
+        for (const auto& [k, v] : config->defaults) {
+            vlog(log.trace, "Default catalog config: '{}': '{}'", k, v);
+            if (k == prefix_prop_name && !v.empty()) {
```
Do we have to check `v.empty()` here?

andrwng replied:
I believe so, I was seeing some empty values coming in.

WillemKauf replied:
Is an empty value a valid default?

andrwng replied:
I don't believe so, given I saw empty values coming in, and they did not correspond go a proper warehouse.

WillemKauf commented on tests/rptest/tests/datalake/datalake_services.py:
```diff
@@ -85,12 +85,8 @@ def setUp(self):
-                "true"
-            })
+            self.redpanda.add_extra_rp_conf(
+                {"iceberg_disable_snapshot_tagging": "true"})
```
> so instead I've updated Nessie tests to not set this property so Nessie can return the default prefix.

Am I reading this correctly that when `iceberg_rest_catalog_warehouse` is specified, Nessie returns `main|main`, and when it is not specified, Nessie returns `main`?

andrwng replied:
Yeah that was the case, and that still is the case. I'm not sure why it was being problematic before, but it appears to work now even when specifying "main" so I've reverted this bit

WillemKauf replied:
So that means Nessie works with requests that have `main|main` in the URI?

andrwng replied:
Right (technically it was main%7Cmain, but yea pretty much)

WillemKauf replied:
Hmm interesting

WillemKauf reviewed (commented):
Looks pretty good, glad we're doing the right thing.

dotnwat commented on src/v/config/configuration.cc:
```diff
@@ -3921,12 +3921,14 @@ configuration::configuration()
       *this,
-      "iceberg_rest_catalog_prefix",
-      "Prefix part of the Iceberg REST catalog URL. Prefix is appended to the "
-      "catalog path f.e. '/v1/{prefix}/namespaces'",
-      {.visibility = visibility::user},
+      "iceberg_rest_catalog_warehouse",
+      "Warehouse to use for the Iceberg REST catalog. Redpanda will query the "
+      "catalog for configurations specific to the warehouse, for example, "
+      "using it to automatically configure the appropriate prefix.",
+      {.visibility = visibility::user,
+       .aliases = {"iceberg_rest_catalog_prefix"}},
```
is the expectation that after upgrade `iceberg_rest_catalog_warehouse` will take on the old/previous value of `iceberg_rest_catalog_prefix`? i don't think that is true (pretty sure i confirmed this recently, but i'm doubting myself now after talking with @IoannisRP about the same thing), but also maybe it's low stakes for this since it would only have been beta enabled previously?

andrwng replied:
Oh hmm I would have expected it to be honored as `iceberg_rest_catalog_warehouse` after upgrade, but I'd believe what you're saying is true.. I think the more important thing is that if a user knows is following old docs and sets `iceberg_rest_catalog_prefix`, then that should be honored and taken as the warehouse.

dotnwat commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -288,6 +350,10 @@ ss::future<expected<create_namespace_response>>
 catalog_client::create_namespace(
   create_namespace_request req, retry_chain_node& rtc) {
     auto gh = _gate.hold();
+    auto config_result = co_await maybe_configure(rtc);
```
i think i only saw trace logging in maybe_configure i wonder if the first successful configure (i'm assuming that happens just once) should announce at info level what happened, or else we see that information some where else?

andrwng replied:
You found em

dotnwat commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -397,23 +479,29 @@ ss::future<expected<commit_table_response>> catalog_client::commit_table_update(
+void path_components::reset_prefix(std::optional<prefix_path> path) {
+    _prefix = trim_slashes<prefix_path>(std::move(path), "");
+    vlog(
+      log.info, "Reset Iceberg REST catalog client prefix to '{}'", *_prefix);
```
oh maybe this is what i was looking for

andrwng replied:
Right exactly, we're just logging at info level once when the configuration happens.

dotnwat commented on src/v/iceberg/rest_client/catalog_client.h:
```diff
@@ -159,6 +167,13 @@ class catalog_client {
+    // Configures this client based on configs from /v1/config in the catalog.
+    // Sets the prefix. Must be called before calls that use the prefix.
+    //
+    // TODO: honor more configs. As implemented, we only support getting the
+    // prefix.
```
but this is all we need to handle the prefix bootstrapping right?

andrwng replied:
That's right, that's the most important one AFAICT.

andrwng committed:
iceberg/rest_client: log more context on errors

I found it really difficult to debug errors coming from the REST
catalog. This commit exposes more information about error responses, for
example:

```
WARN  2025-03-28 14:23:13,021 [shard 2:main] iceberg - catalog_client.cc:332 - [/api/2.1/unity-catalog/iceberg-rest/v1/iceberg_catalog_cdt_devprod_e2e_test_02/namespaces/redpanda/tables/events] error: http_call_error: Bad Request, message: 'Invalid Token'
```

andrwng commented:
Force push to fix a couple tests.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63912-->
<!--2896dfa4-3d30-4f6a-9abd-103b38216eef-->
<!--d803a7cd-9e14-4a8f-93d5-ab1c8b50df2b-->
<!--title-->
## Retry command for [Build#63912](https://buildkite.com/redpanda/redpanda/builds/63912)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/many_topics_test.py::DatalakeManyTopicsTest.test_basic@{"cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a049d2ca-c3a1-4670-b6dc-0bc1f7ddcd08-->
<!--TEST RESULTS-->
<!--acfa6a70-0259-4125-9b73-aeba3eba6142-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63912">build#63912</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63912#0195df7e-50cf-463d-89bf-a9dabb961496 | FLAKY | 1/2 |
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfda-8b75-406f-be26-f4ba7a96466c | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfd7-a666-4f63-a8e5-c89bbb0ddb06 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_many_partitions.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfd7-a667-476a-8d3b-607e1553e6a7 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfd7-a666-4f63-a8e5-c89bbb0ddb06 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfda-8b75-406f-be26-f4ba7a96466c | FLAKY | 1/2 |
| rptest.tests.datalake.many_topics_test.DatalakeManyTopicsTest.test_basic.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfd7-a667-476a-8d3b-607e1553e6a7 | FAIL | 0/20 |
| rptest.tests.datalake.many_topics_test.DatalakeManyTopicsTest.test_basic.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfda-8b73-439e-986e-616f897bfff2 | FAIL | 0/20 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfda-8b76-413f-9189-571f8021ced7 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfda-8b75-406f-be26-f4ba7a96466c | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/63912#0195dfd7-a667-476a-8d3b-607e1553e6a7 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63917">build#63917</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                                                       | unit        | https://buildkite.com/redpanda/redpanda/builds/63917#0195e0e9-a730-4afe-8326-59cc073f3f7f | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63917#0195e12b-c20a-4248-8441-ae96a4724b77 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63917#0195e12b-c209-434c-8133-c9387cf8af97 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63917#0195e12f-1771-4e85-9fd9-9a699c4f6fb7 | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat reviewed (approved):
lgtm. i didn't read through willem's comments

vbotbuildovich commented:
/backport v25.1.x
